kvm/src/lib.rs

```diff
@@ -1004,6 +1013,34 @@ mod tests {
         }
     }
 
+    // Helper function for mmap an anonymous memory of `size`.
+    // Panics if the mmap fails.
+    fn mmap_anonymous(size: usize) -> *mut u8 {
+        let addr = unsafe {
+            libc::mmap(
+                null_mut(),
+                size,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_ANONYMOUS | libc::MAP_SHARED | libc::MAP_NORESERVE,
+                -1,
+                0,
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            panic!("mmap failed.");
+        }
+
+        return addr as *mut u8;
+    }
+
+    impl KvmRunWrapper {
```

Comment:
Given that this is used only in the test that checks the behavior in case of a faulty vcpu fd, it is not necessary to implement another KvmRunWrapper. See my suggestion from the test where this is used.

Reply:
This is not used only in tests. The need for a KvmRunWrapper is detailed in the comment above the structure definition. Specifically, we need a wrapper and we cannot use a raw pointer (defined as either *mut u8 or *mut kvm_run) because when we spawn the vcpu threads the VcpuFd structure is moved. To safely move a structure between threads all the fields of the structure need to implement Send and Sync.

Reply:
> This is not used only in tests. The need for a KvmRunWrapper is detailed in the comment above the structure definition. Specifically, we need a wrapper and we cannot use a raw pointer (defined as either *mut u8 or *mut kvm_run) because when we spawn the vcpu threads the VcpuFd structure is moved. To safely move a structure between threads all the fields of the structure need to implement Send and Sync.

 I see *two* `impl KvmRunWrapper` inside the kvm/src/lib.rs. One that is used for the reasons that you explained and another one used only in tests. Can you please double check?

Reply:
Ah ok, I get it now. So, the `KvmRunWrapper` is not another implementation for `KvmRunWrapper`, it is just an extension of the existing `KvmRunWrapper`. Basically, on top of the existing methods, we also add a new method.

The reason I decided against using a hardcoded address for kvm_run like you did in the example is because the test is checking the errors in case you have a bad file descriptor for either Kvm, VmFd or VcpuFd. If we also pass an invalid pointer (that is not mapped) for kvm_run then we are altering 2 fields and we cannot be sure which one of them generated the errors.